### PR TITLE
Ab#83681 support for context in front office without templates

### PR DESCRIPTION
**Pull request by MwanPygmay**
# Description

Add support for context data by url in front-office (even when no template exists). Created a query to get a temporary dashboard that is a copy of the default one, with added context. Query is executed if no template can be found.
If the data does not exist, we create the temporary dashboard

## Useful links

- [Please insert link to ticket](https://dev.azure.com/WHOHQ/EMSSAFE/_workitems/edit/83681)
- [Please insert link to back-end branch if any](https://github.com/ReliefApplications/ems-backend/pull/946)

## Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] Improvement (refactor or addition to existing functionality)

# How Has This Been Tested?

Created a dashboard with context datasource, two widgets (one for ref data, the other for resource), added some templates (in my case bulbasaur and rattata were templates), and tested. When template exists, it is loaded, if it does not exist we still take into account the context data.

## Screenshots

contextDataInFront

# Checklist:

( * == Mandatory ) 

- [x] * I have set myself as assignee of the pull request
- [x] * My code follows the style guidelines of this project
- [x] * Linting does not generate new warnings
- [x] * I have performed a self-review of my own code
- [x] * I have put the ticket for review, adding the oort-frontend team to the list of reviewers
- [x] * I have commented my code, particularly in hard-to-understand areas
- [x] * I have put JSDoc comment in all required places
- [x] * My changes generate no new warnings
- [x] * I have included screenshots describing my changes if relevant
- [x] * I have selected labels in the Pull Request, according to the changes with code brings
- [ ] I have made corresponding changes to the documentation ( if required )
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by AntoineRelief** [comment]
@matheus-relief 
for review
I'm not 100% sure we need to build a new query for that in the back-end
if we query the dashboard, perhaps we could find an easier way to fetch the context?

**Comment by matheus-relief** [comment]
@AntoineRelief I refactored it a bit. I removed the new query (and another one that we used for creating new contextualized dashboards). Instead, now we can directly query the dashboard with the record id for the context. If there is one it'll be returned, and if there isn't a new one will be created and returned.

**Comment by AntoineRelief** [comment]
@matheus-relief 
one question about your last comment
if you're a front office user, does it mean it will create a new dashboard as well?

**Comment by matheus-relief** [comment]
@MwanPygmay @AntoineRelief Addressed both your comments. Now it only creates a new template if you're on the backoffice. 

The JSDoc I left on the dashboard query sums it up pretty well I think

```
/** Arguments for the dashboard query */
type DashboardArgs = {
  id: string | Types.ObjectId;
  contextEl: any;
  createIfMissing: boolean;
};

/**
 * Throw GraphQL error if not logged.
 * Returns the dashboard by id if no contextEl is provided.
 * If contextEl is provided and its template already exists, returns the template.
 * If contextEl is provided and its template does not exist:
 *   - if createIfMissing is false, returns the main dashboard with the relevant context
 *   - if createIfMissing is true, creates a new template for the element and returns it (if user has permissions)
 */
```

**Comment by RenzoPrats** [comment]
@matheus-relief @MwanPygmay I found some issues:

- When change the select of the context it does not update the dashboard
- Some items of the context returns an error

Peek 24-01-2024 12-06

When we enter in a context edit it, go to another and then come back it seems that it reset
Peek 24-01-2024 12-07

**Review by RenzoPrats — CHANGES_REQUESTED** [review]
@matheus-relief @MwanPygmay I found some issues:

- When change the select of the context it does not update the dashboard
- Some items of the context returns an error

Peek 24-01-2024 12-06

When we enter in a context edit it, go to another and then come back it seems that it reset
Peek 24-01-2024 12-07

**Comment by MwanPygmay** [comment]
@RenzoPrats the ticket is about changes in front-office, I think the tests you are doing are done from back-office, if I am not mistaken. If you found a bug in back-office, please put it in oort-bug channel, and we will address it in a new ticket!

**Comment by AntoineRelief** [comment]
@MwanPygmay 
based on last changes from @matheus-relief I think both are now related

**Comment by RenzoPrats** [comment]
@MwanPygmay that's true, sorry I thought could be about the context of this ticket. So in the front office the issue that I can see is: 
Default dashboard
Captura de tela de 2024-01-24 14-19-45
Configure a dashboard with context in backoffice
Captura de tela de 2024-01-24 14-20-27
When navigate to frontoffice in the default dashboard we can see the chart I configured before but without data:  
Captura de tela de 2024-01-24 14-22-57
and if we go to the one with context it looks like the default one of backoffice.
Captura de tela de 2024-01-24 14-23-10

I'm not sure if this is scope of this ticket, the issue about see the default when we navigate to a dashboard context that is not created yet is working.

**Comment by matheus-relief** [comment]
Thanks @RenzoPrats!! I fixed a couple of the problems:

* ~~When change the select of the context it does not update the dashboard~~
* ~~When we enter in a context edit it, go to another and then come back it seems that it reset~~: The issue there was we were editing the main dashboard instead of the templates

This one I could not replicate:

> Some items of the context returns an error

I even tried the same refData and config but no luck, and I also looked and I couldn't find .replace anywhere in the back on the dashboard query or field resolvers, so I'm not sure where the error is coming from... Can you share the error you were getting?

This one I think it's just a side effect you the bug I fixed where you edited the wrong dashboard. 

> When navigate to frontoffice in the default dashboard we can see the chart I configured before but without data

**Review by RenzoPrats — APPROVED** [review]
Hi @matheus-relief, thanks for the repairs it's working fine, about the issue that I mentioned I found where it was happening in the code and fixed. it was happening when we move to a dashboard with context and create some widget, when we navigate to another id and come back wasn't possible to get the settings of the widget  causing errors. Now I think it's working fine.

File: apps/back-office/src/app/dashboard/pages/dashboard/dashboard.component.ts [hunk 1]
```diff
@@ -247,7 +247,7 @@ export class DashboardComponent
           this.id = data.dashboard.id || id;
           this.dashboard = data.dashboard;
           this.gridOptions = {
-            ...omit(this.gridOptions, ['gridType', 'minimumHeight']), // Prevent issue when gridType or minimumHeight was not set
+            ...omit(this.gridOptions, 'gridType'), // Prevent issue when gridType was not set
```
Comment (AntoineRelief) [inline apps/back-office/src/app/dashboard/pages/dashboard/dashboard.component.ts]: @RenzoPrats 
why removing this?

File: apps/back-office/src/app/dashboard/pages/dashboard/dashboard.component.ts [hunk 2]
```diff
@@ -286,10 +286,6 @@ export class DashboardComponent
           this.buttonActions = this.dashboard.buttons || [];
           this.showFilter = this.dashboard.filter?.show ?? false;
           this.contextService.isFilterEnabled.next(this.showFilter);
-          this.contextService.filterPosition.next({
```
Comment (AntoineRelief) [inline apps/back-office/src/app/dashboard/pages/dashboard/dashboard.component.ts]: @RenzoPrats 
same there

Comment (RenzoPrats) [reply apps/back-office/src/app/dashboard/pages/dashboard/dashboard.component.ts]: Sorry, I just saw it and I'm gonna fix, I think when I pressed ctrl+z I changed to the previous merge branch and caused issue.

**Comment by AntoineRelief** [comment]
missing:
- not anymore a notification when creating a new template
- sometimes the page is just blank for no reason, and I get tons of messages saying that readOnly properties cannot be overwritten

**Comment by AntoineRelief** [comment]
@MwanPygmay @matheus-relief @RenzoPrats 
now seems to work fine!

there was a few things missing:
- context not correctly passed to context service
- widgets were immutable in front-office in some cases
- web-widgets app not updated ( but that, I can't blame you for it )

File: apps/back-office/src/environments/environment.ts [hunk 3]
```diff
@@ -3,29 +3,18 @@ import { theme } from '../themes/default/default.local';
...
   postLogoutRedirectUri: 'http://localhost:4200/auth/',
-  clientId: 'oort-client',
-  scope: 'openid profile email offline_access',
+  clientId: '021202ac-d23b-4757-83e3-f6ecde12266b',
+  scope:
+    'openid profile email offline_access offline_access api://75deca06-ae07-4765-85c0-23e719062833/access_as_user',
+  // Last scope is used to authenticate against Common Services
   responseType: 'code',
   showDebugInformation: true,
+  strictDiscoveryDocumentValidation: false,
```
Comment (matheus-relief) [inline apps/back-office/src/environments/environment.ts:17]: Was this committed on purpose?

**Pull request merged**